### PR TITLE
Assert that HIR nodes are not their own parent.

### LINKED ISSUE #101505: regression: hang with -Zsave-analysis
### Code

I tried this code:

```rust
fn next_letter_position(&self, x_index: f32, y_index: f32, width: f32) -> (f32, f32) {
    let vwidth = &self.viewport.x / 2.0;

    match x_index * width {
        0.0 ... vwidth => {
            (x_index + 1.0, y_index)
        },
        _ => (0.0, y_index - 1.0)
    }
}
```

with rustc `rustc 1.65.0-nightly (c07a8b4e0 2022-08-26)` this runs in a couple ms:
`rustc ./a.rs  --crate-type=lib  -Zsave-analysis  -Ztime=yes -Ztime-passes=yes`

with rustc `rustc 1.65.0-nightly (3c7278846 2022-09-06)` this seems to hang somewhere in `save-analysis`

````
  26,77%  librustc_driver-0f3b2583c1b011ce.so  [.] <rustc_middle::hir::map::Map>::get
  24,00%  librustc_driver-0f3b2583c1b011ce.so  [.] <rustc_middle::hir::map::Map>::get_parent_node
   5,76%  librustc_driver-0f3b2583c1b011ce.so  [.] <rustc_save_analysis::SaveContext>::get_path_res
````
( I killed the process after waiting for 5 minutes :upside_down_face: )

### PR DESCRIPTION
**Pull request by cjgillot**
Fixes https://github.com/rust-lang/rust/issues/101505.
Replaces #101513

r? @petrochenkov @nnethercote

### CODE REVIEW

File: compiler/rustc_middle/src/hir/map/mod.rs [hunk 1]
```diff
@@ -301,6 +301,8 @@ impl<'hir> Map<'hir> {
             let owner = self.tcx.hir_owner_nodes(id.owner).as_owner()?;
             let node = owner.nodes[id.local_id].as_ref()?;
             let hir_id = HirId { owner: id.owner, local_id: node.parent };
+            // HIR indexing should have checked that this.
```
Comment (Rageking8) [inline compiler/rustc_middle/src/hir/map/mod.rs:304]: Is there a typo in this comment? The sentence seems to cut off or has an additional word.

**Comment by petrochenkov** [comment]
Thanks!
r=me with https://github.com/rust-lang/rust/pull/101688#discussion_r967837312 addressed.
@rustbot author

**Comment by nnethercote** [comment]
Thanks, @cjgillot! #101513 added a test, any chance that could be incorporated into this PR?

**Pull request merged**